### DISCUSSION TITLE
How to find out key codes?

### DISCUSSION THREAD

**Discussion by stfnx** [xremap/xremap#356] [Q&A] [answered]
Hi,
my mouse has a horizontal scroll wheel and two buttons on the side. The
wheel I want to remap to volume up/down and the buttons to mute and skip to next
song.

I'm on NixOS Wayland Gnome and found xremap could be a good fit in this
constellation, since it supports Wayland and you even got a Nix Flake,
which is very nice. 😉

But I don't know how I can find the button codes/names.
I tried it with **wev**, which shows the following for the buttons:
```
[13:      wl_pointer] button: serial: 2794; time: 1763135; button: 276
[13:      wl_pointer] button: serial: 3001; time: 1804112; button: 275
```
So the button codes are 275 and 276, but I can't find them in this list:
https://github.com/emberian/evdev/blob/1d020f11b283b0648427a2844b6b980f1a268221/src/scancodes.rs#L26-L572

And for the scroll wheel I have absolutely no idea, wev just shows this:
```
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] frame
[13:      wl_pointer] axis_source: 0 (wheel)
[13:      wl_pointer] axis_stop: axis: 1 (horizontal), discrete: 1
[13:      wl_pointer] axis: time: 1892102; axis: 1 (horizontal), value: 10.000000
```

I'd appreciate your help in this case.

Thank you!

**Comment by k0kubun** [comment]
I'll consider providing alternatives later, but one quick tip is that you can run "RUST_LOG=debug xremap ..." and that gives you the actual key codes it recognizes.

**Reply by stfnx** [reply]
Seems like I can't run it without providing a configuration.
But I can't build a configuration yet, because for that I need to know the button codes.

```
❯ RUST_LOG=debug ./xremap
error: the following required arguments were not provided:
  <CONFIGS>...

Usage: xremap <CONFIGS>...

For more information, try '--help'.
```

**Reply by k0kubun** [reply]
Consider using an empty config, e.g remap:{} under keymap.

**Reply by stfnx** [reply]
I got it to work and could find out the codes for the buttons: `BTN_EXTRA` and `BTN_SIDE`.
But for the scroll wheel its still not clear:
```
# wheel up
[2023-10-07T16:15:07Z DEBUG xremap::event_handler] => 1: unknown key: 59987
[2023-10-07T16:15:07Z DEBUG xremap::event_handler] => 0: unknown key: 59987
[2023-10-07T16:15:07Z DEBUG xremap::event_handler] => 1: unknown key: 59999
[2023-10-07T16:15:07Z DEBUG xremap::event_handler] => 0: unknown key: 59999

# wheel down
[2023-10-07T16:17:32Z DEBUG xremap::event_handler] => 1: unknown key: 59986
[2023-10-07T16:17:32Z DEBUG xremap::event_handler] => 0: unknown key: 59986
[2023-10-07T16:17:32Z DEBUG xremap::event_handler] => 1: unknown key: 59998
[2023-10-07T16:17:32Z DEBUG xremap::event_handler] => 0: unknown key: 59998
```

**Reply by k0kubun** [reply, marked as answer]
Hmm, they are tricky. [These lines](https://github.com/k0kubun/xremap/blob/be51386e42583255dddc2777c77b1a28defea8d4/src/config/key.rs#L86-L106) seem to be relevant to this problem.

Can you try using the following key codes?

* For 59986: `XRIGHTSCROLL`
* For 59987: `XLEFTSCROLL`
* For 59998: `XHIRES_RIGHTSCROLL`
* For 59999: `XHIRES_LEFTSCROLL`

**Reply by stfnx** [reply]
Those work! Thank you!
Seems like it doesn't matter if I use the HIRES version or the other one.
Both variants lead to the same result in this case.

But at the same time my normal scroll wheel stopped working...
I used `RUST_LOG=debug` again to find out that my normal scroll wheel uses the codes `59996` and `59997`. So those are different codes and therefore I'm not sure where this side effect comes from. I assume the normal scroll wheel is `XSCROLLUP` and `XSCROLLDOWN`. I would try rebinding those with too, but I'm not sure to what?

**Reply by k0kubun** [reply]
I think it's a different problem. Can you try using xremap v0.8.10 or newer and pass `--mouse` to it? (I might add a different interface for this in a newer version)

If you've been already using `--mouse`, please let me know. I'll need to come up with something else.

**Reply by stfnx** [reply]
Running xremap with `--mouse` works perfectly. Thanks!
Sadly the Nix Flake doesn't have an option for this yet. I'll create a feature request.

**Reply by GustawXYZ** [reply]
Hey, I've just came across this issue. I've tried remapping the XRIGHTSCROLL buttons to something else, but that doesn't seem to work. They also show up as unknown despite the `--mouse` flag:
```
$ xremap --version
xremap 0.10.7

$ RUST_LOG=debug xremap --mouse ~/config.yml --device "ELECOM TrackBall Mouse DEFT Pro TrackBall"
Selecting devices from the following list:
------------------------------------------------------------------------------
/dev/input/event0 : ELECOM TrackBall Mouse DEFT Pro TrackBall
/dev/input/event1 : ELECOM TrackBall Mouse DEFT Pro TrackBall
/dev/input/event2 : ELECOM TrackBall Mouse DEFT Pro TrackBall Consumer Control
/dev/input/event3 : ELECOM TrackBall Mouse DEFT Pro TrackBall System Control

...

------------------------------------------------------------------------------
Selected devices matching ["ELECOM TrackBall Mouse DEFT Pro TrackBall"]:
------------------------------------------------------------------------------
/dev/input/event0 : ELECOM TrackBall Mouse DEFT Pro TrackBall
/dev/input/event1 : ELECOM TrackBall Mouse DEFT Pro TrackBall
/dev/input/event2 : ELECOM TrackBall Mouse DEFT Pro TrackBall Consumer Control
/dev/input/event3 : ELECOM TrackBall Mouse DEFT Pro TrackBall System Control
------------------------------------------------------------------------------
[2025-03-06T19:55:31Z DEBUG xremap::event_handler] => 1: unknown key: 59987
[2025-03-06T19:55:31Z DEBUG xremap::event_handler] => 0: unknown key: 59987
[2025-03-06T19:55:31Z DEBUG xremap::event_handler] => 1: unknown key: 59999
[2025-03-06T19:55:31Z DEBUG xremap::event_handler] => 0: unknown key: 59999
[2025-03-06T19:55:32Z DEBUG xremap::event_handler] => 1: unknown key: 59986
[2025-03-06T19:55:32Z DEBUG xremap::event_handler] => 0: unknown key: 59986
[2025-03-06T19:55:32Z DEBUG xremap::event_handler] => 1: unknown key: 59998
[2025-03-06T19:55:32Z DEBUG xremap::event_handler] => 0: unknown key: 59998
```
Any suggestions?

**Reply by GustawXYZ** [reply]
Never mind that, remapping seems to actually work, I just forgot `mouse = true;` when testing. It's only the debug output that doesn't seem to recognize those keys still.

**Comment by stfnx** [comment]
I tried to use xev instead of wev and it outputs the following codes. Is this worth anything?
- button_top: 9 
- button_bot: 8 
- wheel_up: 7
- wheel_down: 6

**Comment by Vaisakhkm2625** [comment]
Don't use `xev`/`wev` for finding keys, use `evtest` utility or use xremap in debug mode with `RUST_LOG=debug xremap ~/.config/xremap/config.yml` 

as there is difference between `keycodes` which is reported by `xev/wev` and `scancodes` which is used by `evdev`,`evtest` or `xremap`

**Comment by anasouardini** [comment]
`xev | awk -F'[ )]+' '/^KeyPress/ { a[NR+2] } NR in a { printf "%-3s %s\n", $5, $8 }'`
That shall work.
